idea-plugin/src/main/kotlin/arrow/meta/ide/plugins/quotes/lifecycle/QuoteLifecycle.kt

```diff
@@ -0,0 +1,185 @@
+import com.intellij.openapi.application.Application
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.ModalityState
+import com.intellij.openapi.application.ReadAction
+import com.intellij.openapi.application.runReadAction
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.editor.EditorFactory
+import com.intellij.openapi.editor.event.BulkAwareDocumentListener
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.openapi.progress.EmptyProgressIndicator
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.roots.FileIndexFacade
+import com.intellij.openapi.util.Key
+import com.intellij.openapi.vfs.AsyncFileListener
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.openapi.vfs.VirtualFileManager
+import com.intellij.openapi.vfs.newvfs.events.VFileContentChangeEvent
+import com.intellij.openapi.vfs.newvfs.events.VFileCopyEvent
+import com.intellij.openapi.vfs.newvfs.events.VFileEvent
+import com.intellij.openapi.vfs.newvfs.events.VFileMoveEvent
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.testFramework.LightVirtualFile
+import com.intellij.util.ui.update.Update
+import org.jetbrains.kotlin.idea.debugger.readAction
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+import java.util.concurrent.atomic.AtomicBoolean
+
+data class QuoteConfigs(
+  val quoteSystem: QuoteSystemService,
+  val cache: QuoteCache
+)
+
+fun Project.quoteConfigs(): QuoteConfigs? =
+  getService(QuoteSystemService::class.java)?.let { quoteSystem ->
+    getService(QuoteCache::class.java)?.let { cache ->
+      QuoteConfigs(quoteSystem, cache)
+    }
+  } ?: throw UnavailableServices(listOf(QuoteSystemService::class.java, QuoteCache::class.java))
+
+internal fun quoteProjectOpened(project: Project): Unit = // previously in QuoteSystemCache.projectOpend()
+  project.quoteConfigs()?.let { (quoteSystem: QuoteSystemService, cache: QuoteCache) ->
+    // add a startup activity to populate the cache with a transformation of all project files
+    //StartupManager.getInstance(project).runWhenProjectIsInitialized {
+    //runBackgroundableTask("Arrow Meta", project, cancellable = false) {
+    try {
+      LOG.info("Initializing quote system cache...")
+      val files: List<KtFile> = runReadAction {
+        project.quoteRelevantFiles().also {
+          LOG.info("collected ${it.size} quote relevant files for Project:${project.name}")
+        }
+      }
+      quoteSystem.refreshCache(cache, project, files, analysisIdeExtensions, cacheStrategy())
+    } finally {
+      /*try { // why
+      flushData()
+    } catch (e: Exception) {
+    }*/
+      project.getService(QuoteHighlightingCache::class.java)?.run {
+        map {
+          it.copy(initialized = AtomicBoolean(true))
```

Comment:
This is possibly a bug. This creates a new `AtomicBoolean` instead of updating the existing value with `true`. That's used to allow thread-safe updates of the value. 
Now it's a reassignment to the `PersistentState.value` variable via `map`. `value` isn't volatile, so this may break thread-safety.